dopamine/discrete_domains/run_experiment.py

```diff
@@ -555,10 +570,36 @@ def __init__(
         environment, and returns an agent.
       create_environment_fn: A function which receives a problem name and
         creates a Gym environment for that problem (e.g. an Atari 2600 game).
+      max_simultaneous_iterations: int, maximum number of iterations running
+        simultaneously in separate threads.
       **kwargs: Additional positional arguments.
     """
     threading_utils.initialize_local_attributes(
         self, _environment=create_environment_fn)
+    self._running_iterations = threading.Semaphore(max_simultaneous_iterations)
+    self._output_lock = threading.Lock()
+
     super(AsyncRunner, self).__init__(
         base_dir=base_dir, create_agent_fn=create_agent_fn,
         create_environment_fn=create_environment_fn, **kwargs)
+
+  # TODO(aarg): Decouple experience generation from training.
+  def _run_experiment_loop(self):
+    """Runs iterations in multiple threads until `num_iterations` is reached."""
+    threads = []
+    for iteration in range(self._start_iteration, self._num_iterations):
+      self._running_iterations.acquire()
+      threads.append(self._run_one_iteration(iteration))
+    # Wait for all running iterations to complete.
+    for thread in threads:
+      thread.join()
+
+  @async_method
+  def _run_one_iteration(self, iteration):
```

Comment:
once a trajectory is finished - it is stored to the main buffer from which they are going to be sampled at any training step.
The length of an episode is typically much shorter than the buffer size hence it should no matter that the storage of an episode is delayed (until the end of the episode is reached).

Reply:
According to my thinking, the model can only be fit to a batch of experiences that have been saved to the main replay buffer.  Will this mean many train steps on an replay buffer that changes only occasionally?

Reply:
```According to my thinking, the model can only be fit to a batch of experiences that have been saved to the main replay buffer```
-> that's correct

```Will this mean many train steps on an replay buffer that changes only occasionally``` the replay buffer is updated frequently (originally at every step, in the async implementation at each episode). Though it changes slowly because the size of the buffer is much larger than the size on an episode.

Reply:
>> Will this mean many train steps on an replay buffer that changes only occasionally?

> The replay buffer is updated frequently (originally at every step, in the async implementation at each episode)

How can this be?  In the async, `use_contiguous_trajectories` context, the global replay buffer is only updated when an environment reaches a terminal step.

Reply:
Yes we agree, the replay buffer is update at every episode.

Do you see it as an issue? The size of the buffer is typically much larger than episode length's so this difference in update frequency shouldn't matter too much - am I wrong?

Reply:
I think you're right.  It's a good point.  I saw the word "frequently" and got tunnel vision -- sorry for that.

Resolving.